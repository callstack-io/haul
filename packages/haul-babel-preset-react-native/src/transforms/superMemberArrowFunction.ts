import createTemplate from '@babel/template';
import * as babel from '@babel/core';

/**
 * With Webpack and babel-loader the following code:
 * ```js
 * class Derived extends Super {
 *   derivedFn1 = async () => {
 *     await super.superFn1();
 *   }
 *
 *   derivedFn2 = () => {
 *     super.superFn2();
 *   }
 * }
 * ```
 * will fail with an exception: `super` cannot be used outside of a class method.
 * However this is a valid syntax, which successfully run for example on V8 or Chrome,
 * so we need to transform `super` to use `Super.prototype`:
 * ```js
 * class Derived extends Super {
 *   derivedFn1 = async () => {
 *     await Super.prototype.superFn1.call(this);
 *   }
 *
 *   derivedFn2 = () => {
 *     Super.prototype.superFn2.call(this);
 *   }
 * }
 * ```
 */

const template = createTemplate('SUPER.prototype.METHOD.call(this, ARGS)');

export default function superMemberArrowFunction() {
  return {
    visitor: {
      Program(programPath: babel.NodePath<babel.types.Program>) {
        programPath.traverse({
          Super(path) {
            let parentFunctionPath = path.getFunctionParent();

            // Check if parent function is an arrow function as a class property.
            if (
              parentFunctionPath.isArrowFunctionExpression() &&
              parentFunctionPath.parentPath.isClassProperty()
            ) {
              // Find the parent class, and make sure it's extending from an identifier
              let parentClass = path.findParent(p =>
                p.isClass()
              ) as babel.NodePath<babel.types.Class>;

              if (
                parentClass?.node.superClass &&
                babel.types.isIdentifier(parentClass.node.superClass)
              ) {
                path.parentPath.parentPath.replaceWith(
                  template({
                    SUPER: parentClass.node.superClass,
                    METHOD: (path.parent as babel.types.MemberExpression)
                      .property,
                    ARGS: (path.parentPath.parent as babel.types.CallExpression)
                      .arguments,
                  }) as any
                );
              }
            }
          },
        });
      },
    },
  };
}
