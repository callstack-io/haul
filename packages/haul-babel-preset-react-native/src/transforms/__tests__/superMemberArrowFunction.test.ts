import superMemberArrowFunction from '../superMemberArrowFunction';
// @ts-ignore
import classProperties from '@babel/plugin-proposal-class-properties';
import { transformSync } from '@babel/core';

const fixture = `
class Derived extends Super {
  derivedFn1 = async () => {
    await super.superFn1();
  };

  derivedFn2 = () => {
    super.superFn2(123, 456);
  };

  async derivedFn3() {
    await super.superFn3();
  }

  derivedFn4() {
    super.superFn4();
  }
}
`;

describe('superMemberArrowFunction transform', () => {
  it('should replace `super` with <Super>.prototype', () => {
    const { code } = transformSync(fixture, {
      plugins: [superMemberArrowFunction, classProperties],
      babelrc: false,
      configFile: false,
    }) || { code: '' };

    expect(code).toMatch('await Super.prototype.superFn1.call(this);');
    expect(code).toMatch('Super.prototype.superFn2.call(this, 123, 456);');
    expect(code).toMatch('await super.superFn3()');
    expect(code).toMatch('super.superFn4()');
    expect(code).toMatchSnapshot();
  });
});
