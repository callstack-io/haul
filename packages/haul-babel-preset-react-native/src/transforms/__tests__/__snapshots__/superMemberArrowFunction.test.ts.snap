// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`superMemberArrowFunction transform should replace \`super\` with <Super>.prototype 1`] = `
"function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Derived extends Super {
  constructor(...args) {
    super(...args);

    _defineProperty(this, \\"derivedFn1\\", async () => {
      await Super.prototype.superFn1.call(this);
    });

    _defineProperty(this, \\"derivedFn2\\", () => {
      Super.prototype.superFn2.call(this, 123, 456);
    });
  }

  async derivedFn3() {
    await super.superFn3();
  }

  derivedFn4() {
    super.superFn4();
  }

}"
`;
